csi/src/freezefs.rs

```diff
@@ -0,0 +1,75 @@
+//! The files system freeze support using linux utility fsfreeze
+use crate::{
+    dev::{Device, DeviceError},
+    mount,
+};
+use snafu::{ResultExt, Snafu};
+use tokio::process::Command;
+use uuid::Uuid;
+
+#[derive(Debug, Snafu)]
+#[snafu(visibility = "pub(crate)")]
+pub enum FreezeFsError {
+    #[snafu(display("Cannot find volume: volume ID: {}", volid))]
+    VolumeNotFound { volid: String },
+    #[snafu(display("Invalid volume ID: {}, {}", volid, source))]
+    InvalidVolumeId {
+        source: uuid::parser::ParseError,
+        volid: String,
+    },
+    #[snafu(display("fsfreeze failed: volume ID: {}, {}", volid, error))]
+    FsfreezeFailed { volid: String, error: String },
+    #[snafu(display("Internal failure: volume ID:{}, {}", volid, source))]
+    InternalFailure { source: DeviceError, volid: String },
+    #[snafu(display("IO error: volume ID: {}, {}", volid, source))]
+    IOError {
+        source: std::io::Error,
+        volid: String,
+    },
+}
+
+const FSFREEZE: &str = "fsfreeze";
+
+async fn fsfreeze(
+    volume_id: &str,
+    freeze_op: &str,
+) -> Result<(), FreezeFsError> {
+    let uuid = Uuid::parse_str(volume_id).context(InvalidVolumeId {
+        volid: volume_id.to_string(),
+    })?;
+
+    if let Some(device) =
+        Device::lookup(&uuid).await.context(InternalFailure {
```

Comment:
is device::lookup returning an Option<Result<T,E>>? looks to me we want to fix that (create an issue for it?)

Reply:
Yes, the function is defined as `pub async fn lookup( uuid: &Uuid,) -> Result<Option<Box<dyn Detach>>, DeviceError>`

Reply:
This is correct. It is possible that the lookup will fail because of an error. Or it may just fail to find what we're looking for because it is not there - which is not an error.